Migrate yaml e2e test from knative/build

shashwathi opened the pull request:
Fixes https://github.com/knative/build-pipeline/issues/302

Add yaml tests which cover different features of knative build in knative build pipeline like mounting secret as volume, templating arguments.
I followed the pattern suggested by Jason to include relevant k8s, knative objects in same yaml separated by `---`.  Test file names should sufficiently explain the taskrun but if anybody feels strongly about adding more docs, please leave comment on what would be ideal place for this documentation.

I have not intentionally added all tests because either feature is not supported(like failure labels) or better examples are already present in repo like task template params.

cc @ImJasonH

bobcatfish commented on examples/README.md:
```diff
@@ -105,13 +105,30 @@ The [run](./run/) directory contains an example
 [Knative flavored credentials](https://github.com/knative/docs/blob/master/build/auth.md#guiding-credential-selection)
 are supported too.
+
+## Results
```
ooo nice! maybe put a link to this section at the top?

bobcatfish replied:
just bumping this :sweat_smile:

shashwathi replied:
README is pretty small so is there any value in having link at top?

bobcatfish commented on examples/run/task-volume-args.yaml:
```diff
@@ -0,0 +1,42 @@
+metadata:
+  name: config-for-testing
+data:
+  test.data: builds are my jam
```
XD

bobcatfish replied:
"tasks are my jam"? XD XD XD

bobcatfish commented on examples/run/task-env.yaml:
```diff
@@ -0,0 +1,61 @@
+      description: FOODBAR variable
+      default: foobar
+  steps:
+  # Test the environment variables are set in the build
```
there are a few mentions of `build` in the examples that i think would be a bit confusing if you saw the example without understanding the Build -> Task lineage, maybe we can change these to `Task`?

bobcatfish commented on examples/run/taskrun-docker-basic.yaml:
```diff
@@ -0,0 +1,49 @@
+    build.knative.dev/docker-1: https://eu.gcr.io
+    build.knative.dev/docker-2: https://asia.gcr.io
+    build.knative.dev/docker-3: https://gcr.io
+    build.knative.dev/docker-4: https://reduce-chance-of-selecting-gcr.io
```
i think these should be something else, maybe [`pipeline.knative.dev`](https://github.com/knative/build-pipeline/blob/master/docs/auth.md#guiding-credential-selection) ? (not sure if this actually works tho 😅 )

shashwathi replied:
It does not work. Is this something we need a separate issue or tackle it in this PR?

bobcatfish replied:
oh yeah separate PR then for sure i think! separate issue even

shashwathi commented:
@bobcatfish 

>  Would it be a crazy amount of work to include in the commit message which ones were skipped and why? Makes it easy to make sure none were skipped that shouldn't be, and also I can see us having to compare the examples b/w the repos again a couple more times as we migrate development completely away from knative/build

Done. Updated commit msg to include yamls that were not migrated

> * There are a few references to "build" and/or "build crd" in the examples that we might want to update

I think I have addressed it all. Let me know if I missed something

> * What do you think about getting rid of the `run` directory all together, having all the examples in the top level `examples` dir, one file per complete example? I think it's a bit unclear at the moment why some of the examples live at the top level and get special explanation in the README and some don'

I think there are couple of cases like replacing image path which needs section in README but I agree rest of them don't need it I guess.

shashwathi commented:
/hold

e2e yaml bash script does not seem to recognize new files

shashwathi commented:
/hold cancel

shashwathi commented:
@bobcatfish : one of build yaml tests relied on volume source templating and that feature has not implemented yet. 
I have added that support in separate commit from yaml tests in this PR(also unit tests). I can do another PR too but build pipeline merges PR retaining commits so I don't think it makes much difference. If you feel otherwise I can do another PR too.

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/resources/apply.go:
```diff
@@ -89,5 +89,14 @@ func ApplyReplacements(build *buildv1alpha1.Build, replacements map[string]strin
 		}
 	}
+
+	// Apply variable expansion to the build's volumes
```
neat!!

bobcatfish commented:
Generally looking great but a couple doc requests:

- Explaining the structure of the `examples/` dir in `examples/README.md` (at the moment this readme kind of gives the impression that "simple pipeline" and "pipeline with output" are the only examples (I'm still not sure it makes sense to have docs in the main README on only these 2 examples, I could also see moving these descriptions to comments in the yamls themselves)
- Updating the templating docs that we support volumes now too

> Few tasks are referenced in taskruns and pipelinerun so there will be copies of same tasks repeated in multiple files. is that okay?

Hm I could go either way!

If we decided to not duplicate them, I think `examples/README.md`should clearly explain the structure of the dirs, i.e. how the top level files in `examples/` are used across runs.

> I have added that support in separate commit from yaml tests in this PR(also unit tests). I can do another PR too but build pipeline merges PR retaining commits so I don't think it makes much difference. If you feel otherwise I can do another PR too.

I'm fine with same PR but can we add some docs on this to our templating docs?

shashwathi commented:
> I'm fine with same PR but can we add some docs on this to our templating docs?

Done. Added an example. Let me know what you think

> Hm I could go either way!

I have duplicated tasks and made them all self contained. README contains instruction on how to apply, what to change in yaml files and how to check results. Let me know if I missed anything.

shashwathi commented:
@bobcatfish : https://github.com/knative/build-pipeline/issues/511

shashwathi commented:
/test pull-knative-build-pipeline-integration-tests

bobcatfish commented:
Really nice work @shashwathi !!! The examples dir is looking great :D :D :D 

Thanks for catching and fixing those issues we'd missed as well :heart:  - and thanks for taking this on, hope it wasn't too tedious :sweat_smile: 

/lgtm
/meow space
